tests/unit/processing/test_process_download.py

```diff
@@ -75,7 +105,87 @@ def test_download_unknown_uri(self):
         """
 
         uri = "fake://fake_bucket_name/fake_key/fake_file.h5ad"
-        pdv = ProcessDownload(self.business_logic, self.uri_provider, self.s3_provider)
+        pdv = ProcessDownload(Mock(), self.uri_provider, Mock())
         pdv.download_from_s3 = Mock()
         with pytest.raises(ValueError, match=f"Malformed source URI: {uri}"):
             pdv.download_from_source_uri(uri, "fake_local_path")
+
+
+@pytest.fixture
+def mock_ProcessDownload():
+    return ProcessDownload(Mock(), Mock(), Mock())
+
+
+def sample_adata(n_obs: int, n_vars: int):
+    # Create a sample AnnData object for testing
+    adata = MagicMock(spec=scanpy.AnnData, n_obs=n_obs, n_vars=n_vars)
+    return adata
+
+
+@pytest.fixture
+def mock_read_h5ad():
+    with patch("scanpy.read_h5ad") as mock_read_h5ad:
+        mock_read_h5ad.return_value = sample_adata(1, 2 * GB)
+        yield mock_read_h5ad
+
+
+# Arrange
+@pytest.mark.parametrize(
+    "adata, memory_modifier, expected",
+    [
+        (sample_adata(1, 2 * GB), 1, {"Vcpus": 1, "Memory": 4000, "MaxSwap": 0}),  # minimum memory
+        (sample_adata(1, 5 * GB), 1, {"Vcpus": 2, "Memory": 5120, "MaxSwap": 0}),  # above minimum memory
+        (sample_adata(1, 5 * GB), 1.5, {"Vcpus": 2, "Memory": 7680, "MaxSwap": 0}),  # modifier adjusted
+        (sample_adata(1, 64 * GB), 1, {"Vcpus": 16, "Memory": 64000, "MaxSwap": 300000}),  # maximum memory
+    ],
+)
+def test_estimate_resource_requirements_positive(mock_ProcessDownload, adata, memory_modifier, expected):
+    # Act & Assert
+    assert expected == mock_ProcessDownload.estimate_resource_requirements(adata, memory_modifier=memory_modifier)
+
+
+@pytest.mark.parametrize(
+    "environ,expected",
+    [
+        ({"REMOTE_DEV_PREFIX": "/stack/", "DEPLOYMENT_STAGE": "test"}, "dp-test-stack-ingest-process-fake_dataset_id"),
+        ({"REMOTE_DEV_PREFIX": "stack", "DEPLOYMENT_STAGE": "test"}, "dp-test-stack-ingest-process-fake_dataset_id"),
+        ({"DEPLOYMENT_STAGE": "test"}, "dp-test-ingest-process-fake_dataset_id"),
+    ],
+)
```

Comment:
So this parametrize is used to pass the first element from a given tuple in as the first parameter... which is defined as the first element in a comma-delineated str... so `{"REMOTE_DEV_PREFIX": "/stack/", "DEPLOYMENT_STAGE": "test"}` is assigned to `environ`... and `"dp-test-stack-ingest-process-fake_dataset_id"` is `expected`. 

I honestly don't really see what this gets us, other than complexity in the form of adding more function decorators. If anything this is more hacky than inlining these parameters in the function itself.

Reply:
It is funky for sure, and I don’t agree with all of pytest syntax. Sometimes there is too much magic, but if you trust the magic it makes things easier. Also using pytest syntax integrates better with the plugins we use for pytest.

The fact that the tests are short and to the point makes it easier to reason about what specifically is being tested.

Also pytest doesn't play will with python unittest subtests so using parameterize is what we have.

Reply:
`parameterize` is a general implement that is useful to name test cases. So I think you could include a name for the test case (the input and expected output) and you when you run your tests on the command line, you will see something like `test-case-1` (if that is the name of the test case) failed.

This is more interpretable than including the all the test cases in a data structure like a `list` and looping through it and executing the test. A real run here:

```
tests/unit/backend/wmg/api/common/test_wmg_api_helpers.py::test_sanitize_api_query_dict[input_query0-expected_sanitized_query0] PASSED                                                                                                    [ 81%]
tests/unit/backend/wmg/api/common/test_wmg_api_helpers.py::test_sanitize_api_query_dict[input_query1-expected_sanitized_query1] PASSED                                                                                                    [ 83%]
tests/unit/backend/wmg/api/common/test_wmg_api_helpers.py::test_sanitize_api_query_dict[input_query2-expected_sanitized_query2] PASSED                                                                                                    [ 85%]
tests/unit/backend/wmg/api/common/test_wmg_api_helpers.py::test_sanitize_api_query_dict[input_query3-expected_sanitized_query3] PASSED                                                                                                    [ 87%]
tests/unit/backend/wmg/api/common/test_wmg_api_helpers.py::test_sanitize_api_query_dict[input_query4-expected_sanitized_query4] PASSED
```

Reply:
@prathapsridharan given that, how would you rearrange this particular test?